### PR TITLE
Compatibility for auth handlers without refresh (Revisited)

### PR DESCRIPTION
**Pull request by timgerstel**
Old authentication handlers which do not support session refreshing will not longer return a 400 or 500 error due to lack of capability, but will allow the session to expire naturally (user logs out/timeout).

Must be merged with:
https://github.com/zowe/zss-auth/pull/19
https://github.com/zowe/zosmf-auth/pull/14

Signed-off-by: Timothy Gerstel tgerstel@rocketsoftware.com

### CODE REVIEW

**Comment by timgerstel** [comment]
You are correct, good catch.  Will fix then comment back here, thank you @jordanfilteau1995

**Comment by timgerstel** [comment]
Added fix @jordanfilteau1995

File: lib/webauth.js [hunk 1]
```diff
@@ -185,24 +185,39 @@ module.exports = function(authManager) {
...
+          The logic for performing refreshStatus()/authenticate() calls has been wrapped in a check for the presence of
+          the getCapabilities() function because functions that do not include getCapabilities() likely dont support refreshing (or haven't been maintained).  
+          If a service calls refreshStatus on an auth handler that does not support it,
+          { success: true } is returned so that the session may expire naturally and users are not forced to log in every time they refresh.
+        */
+        if(hasGetCapabilities){
+          if (type === SESSION_ACTION_TYPE_REFRESH
+              && authManager.sessionTimeoutMs !== TIMEOUT_VALUE_NO_EXPIRE
+              && (!timeout || timeout < Date.now())) {
+            req.session.zlux = undefined;
```
Comment (1000TurquoisePogs) [inline lib/webauth.js]: This should not be inside of hasGetCapabilities - this is checking if the server itself has timed out, not if an auth plugin has.

Consider: 
if (this stuff) { 
REASON_ZLUX_SESSION_EXPIRE
} else if (hasGetCapabilities && canRefresh) { 
do refresh
} else {
success = true
}

File: lib/webauth.js [hunk 2]
```diff
@@ -185,24 +185,40 @@ module.exports = function(authManager) {
...
             && authManager.sessionTimeoutMs !== TIMEOUT_VALUE_NO_EXPIRE
             && (!timeout || timeout < Date.now())) {
           req.session.zlux = undefined;
           handlerResult = {success:false, reason:REASON_ZLUX_SESSION_EXPIRE};
+        } else if(hasGetCapabilities && handler.getCapabilities().canRefresh === true && functionName == 'refreshStatus'){
+          handlerResult = yield handler[functionName](req, authPluginSession);
         } else {
-          handlerResult = yield handler[functionName](req,
-                                                      authPluginSession);
+          handlerResult = (functionName == 'refreshStatus') ? { success: true } : yield handler[functionName](req, authPluginSession);
```
Comment (fkovinAtRocket) [inline lib/webauth.js]: Line longer than 120 character

File: lib/webauth.js [hunk 3]
```diff
@@ -185,24 +185,40 @@ module.exports = function(authManager) {
...
+       if(!hasGetCapabilities){
+         authLogger.warn(`${pluginID}: getCapabilities() is not a function`);
+       }
+       
+       if (type === SESSION_ACTION_TYPE_REFRESH
             && authManager.sessionTimeoutMs !== TIMEOUT_VALUE_NO_EXPIRE
             && (!timeout || timeout < Date.now())) {
           req.session.zlux = undefined;
           handlerResult = {success:false, reason:REASON_ZLUX_SESSION_EXPIRE};
+        } else if(hasGetCapabilities && handler.getCapabilities().canRefresh === true && functionName == 'refreshStatus'){
```
Comment (fkovinAtRocket) [inline lib/webauth.js]: Line longer than 120 character

Comment (timgerstel) [reply lib/webauth.js]: Resolved @fkovinAtRocket

File: lib/webauth.js [hunk 4]
```diff
@@ -185,24 +185,38 @@ module.exports = function(authManager) {
...
+        const authHandlerStatus = handler.getStatus(authPluginSession);
+        const wasAuthenticated = authHandlerStatus.authenticated;
+        const hasGetCapabilities = !!((typeof handler.getCapabilities) === 'function');
+        let handlerResult;
+
+       if(!hasGetCapabilities){
+         authLogger.warn(`${pluginID}: getCapabilities() is not a function`);
+       }
+
+       if (type === SESSION_ACTION_TYPE_REFRESH
```
Comment (fkovinAtRocket) [inline lib/webauth.js]: This is technically correct, and also probably  this is an attempt to make a minimal change, but in order to make the intent clearer, I would make a bigger change:
1. Get rid of the `functionName` variable completely from this function: we only need `type` to track what's going on, no need to use two variables for that
2. rewrite it somehow like this: (NOT TESTED!!! review carefully)
```
       if (type === SESSION_ACTION_TYPE_REFRESH) {
         if (authManager.sessionTimeoutMs !== TIMEOUT_VALUE_NO_EXPIRE
            && (!timeout || timeout < Date.now())) {
           req.session.zlux = undefined;
           handlerResult = {success:false, reason:REASON_ZLUX_SESSION_EXPIRE};
         } else if (hasGetCapabilities && handler.getCapabilities().canRefresh) {
           handlerResult = yield handler.refreshStatus(req, authPluginSession);
         } else {
           handlerResult = { success: true };
         }        
       } else if (type === SESSION_ACTION_TYPE_AUTHENTICATE) {
         handlerResult = yield handler.authenticate(req, authPluginSession);
       } 
```
In my view, this is more straightforwardly presenting what we want to do here.

Comment (timgerstel) [reply lib/webauth.js]: Reviewed, implemented and tested.  Works perfectly @fkovinAtRocket

**Pull request merged**